ouroboros-network/test/Test/Mux.hs

```diff
@@ -97,52 +150,117 @@ instance Show DummyPayload where
-       return $ DummyPayload blob
+        n <- choose (1, 128)
+        len <- oneof [ return n
+                     , return $ defaultMiniProtocolLimit - n - cborOverhead
+                     , choose (1, defaultMiniProtocolLimit - cborOverhead)
+                     ]
+        p <- arbitrary
+        let blob = BL.replicate len p
+        return $ DummyPayload blob
+      where
+        cborOverhead = 7 -- XXX Bytes needed by CBOR to encode the dummy payload
 
 instance Serialise DummyPayload where
     encode a = CBOR.encodeBytes (BL.toStrict $ unDummyPayload a)
     decode = DummyPayload . BL.fromStrict <$> CBOR.decodeBytes
 
 data InvalidSDU = InvalidSDU {
-      isTimestamp :: !Mx.RemoteClockModel
-    , isIdAndMode :: !Word16
-    , isLength    :: !Word16
+      isTimestamp  :: !Mx.RemoteClockModel
+    , isIdAndMode  :: !Word16
+    , isLength     :: !Word16
+    , isRealLength :: !Int64
     }
 
 instance Show InvalidSDU where
-    show a = printf "InvalidSDU 0x%08x 0x%04x 0x%04x\n"
+    show a = printf "InvalidSDU 0x%08x 0x%04x 0x%04x 0x%04x\n"
                     (Mx.unRemoteClockModel $ isTimestamp a)
                     (isIdAndMode a)
                     (isLength a)
-
-instance Arbitrary InvalidSDU where
-    arbitrary = do
-        ts  <- arbitrary
-        mid <- choose (5, 0xffff) -- TxSubmission with 4 is the highest valid mid
-        len <- arbitrary
-
-        return $ InvalidSDU (Mx.RemoteClockModel ts) mid len
-
-
-startMuxSTM :: (Ord ptcl, Enum ptcl, Bounded ptcl, Mx.ProtocolEnum ptcl)
-            => Mx.MiniProtocolDescriptions ptcl IO
-            -> TBQueue IO BL.ByteString
-            -> TBQueue IO BL.ByteString
+                    (isRealLength a)
+
+data ArbitrarySDU = ArbitraryInvalidSDU InvalidSDU Mx.MuxBearerState Mx.MuxErrorType
+                  | ArbitraryValidSDU DummyPayload Mx.MuxBearerState (Maybe Mx.MuxErrorType)
+                  deriving Show
+
+instance Arbitrary ArbitrarySDU where
+    arbitrary = oneof [ unknownMiniProtocol
+                      , invalidLenght
+                      , validSdu
+                      , tooLargeSdu
+                      ]
+      where
+        validSdu = do
+            b <- arbitrary
+
+            -- Valid SDUs before version negotiation does only make sense for single SDUs.
+            state <- if BL.length (unDummyPayload b) < 0xffff
+                         then arbitrary
+                         else return Mx.Mature
+            let err_m = if state == Mx.Larval || state == Mx.Connected
+                            then Just Mx.MuxUnknownMiniProtocol
+                            else Nothing
+
+            return $ ArbitraryValidSDU b state err_m
+
+        tooLargeSdu = do
+            l <- choose (1 + smallMiniProtocolLimit , 2 * smallMiniProtocolLimit)
+            p <- arbitrary
+            let b = DummyPayload $ BL.replicate l p
+
+            return $ ArbitraryValidSDU b Mx.Mature (Just Mx.MuxIngressQueueOverRun)
```

Comment:
Hmm, this is slightly misleading since it's invalid sdu at the end.

Reply:
I agree that it is slightly misleading. The fault comes from the fact that it is sent as a single message which doesn't give the miniprotocol thread a chance to consume the part of the message. If the same message where chopped into 4 smaller pieces and sent of one by one Mux would be unable to detect the violation.
SDUs for general miniprotocols would cause a failure during version negotiation but they are also valid. Therefor I think it is acceptable to mark this SDU as valid.